### PR TITLE
GH-35969: [Swift] use ArrowType instead of ArrowType.info and add binary, time32 and time64 types

### LINKED ISSUE #35969: [Swift] Update stored type info to use ArrowType from ArrowType.info.  Add binary, time32, and time64
### Describe the enhancement requested

- Update swift classes to use the ArrowType instead of the ArrowType.info enum to store the type information.  This was required due to the time32 and time64 classes requiring to carry more information than just the type (seconds, microseconds, etc..)
- Additional types added are binary, time32, time64 and decimal 128 (uses built in swift decimal type). 

### Component(s)

Swift

### PR DESCRIPTION
**Pull request by abandy**
Changes:
- Updated code to use ArrowType instead of ArrowType.info to store the arrow type in classes and structs.  This was done due to Time32 and Time64 requiring additional information beyond just the basic enum info.
- Add binary, time32 and time64 types.
* Closes: #35969

### CODE REVIEW

File: swift/Arrow/Sources/Arrow/ArrowArray.swift [hunk 1]
```diff
@@ -182,3 +182,52 @@ public class Date64Array: ArrowArray<Date> {
...
+public class BinaryArray: ArrowArray<Data> {
+    public struct Options {
+        public var printAsHex = false
+        public var printEncoding: String.Encoding = .utf8;
+    }
+    
+    public var options = Options()
+    
+    public override subscript(_ index: UInt) -> Data? {
+        get{
```
Comment (kou) [inline swift/Arrow/Sources/Arrow/ArrowArray.swift:198]:
```suggestion
        get {
```

File: swift/Arrow/Sources/Arrow/ArrowArray.swift [hunk 2]
```diff
@@ -182,3 +182,52 @@ public class Date64Array: ArrowArray<Date> {
...
+        }
+    }
+    
+    public override func asString(_ index: UInt) -> String {
+        if self[index] == nil {
+            return ""
+        }
+        
+        let data = self[index]!
+        if(options.printAsHex) {
```
Comment (kou) [inline swift/Arrow/Sources/Arrow/ArrowArray.swift:227]:
```suggestion
        if options.printAsHex {
```

File: swift/Arrow/Sources/Arrow/ArrowType.swift [hunk 3]
```diff
@@ -116,17 +154,21 @@ public class ArrowType {
             return ArrowType.ArrowUnknown
         }
     }
+    
+
```
Comment (kou) [inline swift/Arrow/Sources/Arrow/ArrowType.swift:158]:
```suggestion
```

File: swift/Arrow/Sources/Arrow/ArrowWriterHelper.swift [hunk 4]
```diff
@@ -34,16 +35,21 @@ func toFBTypeEnum(_ infoType: ArrowType.Info) -> Result<org_apache_arrow_flatbuf
...
     } else if infoType == ArrowType.ArrowBool {
         return .success(org_apache_arrow_flatbuf_Type_.bool)
     } else if infoType == ArrowType.ArrowDate32 || infoType == ArrowType.ArrowDate64 {
         return .success(org_apache_arrow_flatbuf_Type_.date)
+    } else if infoType == ArrowType.ArrowTime32 || infoType == ArrowType.ArrowTime64 {
+        return .success(org_apache_arrow_flatbuf_Type_.time)
     }
-
-    return .failure(.unknownType)
+
```
Comment (kou) [inline swift/Arrow/Sources/Arrow/ArrowWriterHelper.swift:47]:
```suggestion

```

File: swift/Arrow/Tests/ArrowTests/ArrayTests.swift [hunk 5]
```diff
@@ -125,6 +125,62 @@ final class ArrayTests: XCTestCase {
...
+                XCTAssertEqual(try binaryArray.isNull(i), true)
+            } else {
+                let stringData = String(bytes: binaryArray[i]!, encoding: .utf8)
+                XCTAssertEqual(stringData, "test" + String(i))
+            }
+        }
+    }
+    
+    func testTime32Array() throws {
+        let time32Builder = try ArrowArrayBuilders.loadTime32ArrayBuilder(.Milliseconds);
```
Comment (kou) [inline swift/Arrow/Tests/ArrowTests/ArrayTests.swift:156]: We may want to add a test for `.Seconds` too.

File: swift/Arrow/Tests/ArrowTests/ArrayTests.swift [hunk 6]
```diff
@@ -125,6 +125,62 @@ final class ArrayTests: XCTestCase {
...
+        let time32Array = try time32Builder.finish()
+        let time32Type = time32Array.arrowData.type as! ArrowTypeTime32
+        XCTAssertEqual(time32Type.unit, .Milliseconds)
+        XCTAssertEqual(time32Array.length, 3)
+        XCTAssertEqual(time32Array[1], 1000000)
+        XCTAssertEqual(time32Array[2], nil)
+    }
+    
+    func testTime64Array() throws {
+        let time32Builder = try ArrowArrayBuilders.loadTime64ArrayBuilder(.Nanoseconds);
```
Comment (kou) [inline swift/Arrow/Tests/ArrowTests/ArrayTests.swift:172]: We may want to add a test for `.Microseconds` too.

Comment (kou) [inline swift/Arrow/Tests/ArrowTests/ArrayTests.swift:172]:
```suggestion
        let time64Builder = try ArrowArrayBuilders.loadTime64ArrayBuilder(.Nanoseconds);
```

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift [hunk 7]
```diff
@@ -66,21 +66,21 @@ final class IPCFileReaderTests: XCTestCase {
...
         switch result {
-        case .success(let rbBatches):
-            recordBatches = rbBatches
+        case .success(let result):
+            recordBatches = result.batches
         case .failure(let error):
             throw error
         }
-
+
```
Comment (kou) [inline swift/Arrow/Tests/ArrowTests/IPCTests.swift:74]:
```suggestion

```

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift [hunk 8]
```diff
@@ -108,33 +108,34 @@ final class IPCFileReaderTests: XCTestCase {
...
+        let writerInfo = ArrowWriter.Info(.recordbatch, schema: fileRBs[0].schema, batches: fileRBs)
+        switch arrowWriter.toStream(writerInfo) {
         case .success(let writeData):
             //read stream back into recordbatches
             try checkBoolRecordBatch(arrowReader.fromStream(writeData))
         case .failure(let error):
             throw error
         }
-
+
```
Comment (kou) [inline swift/Arrow/Tests/ArrowTests/IPCTests.swift:119]:
```suggestion

```

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift [hunk 9]
```diff
@@ -108,33 +108,34 @@ final class IPCFileReaderTests: XCTestCase {
...
+        switch arrowWriter.toFile(outputUrl, info: writerInfo) {
         case .success(_):
             try checkBoolRecordBatch(arrowReader.fromFile(outputUrl))
         case .failure(let error):
             throw error
         }
         
     }
-
+
```
Comment (kou) [inline swift/Arrow/Tests/ArrowTests/IPCTests.swift:130]:
```suggestion

```

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift [hunk 10]
```diff
@@ -108,33 +108,34 @@ final class IPCFileReaderTests: XCTestCase {
...
-        return schemaBuilder.addField("col1", type: ArrowType.ArrowUInt8, isNullable: true)
-            .addField("col2", type: ArrowType.ArrowString, isNullable: false)
-            .addField("col3", type: ArrowType.ArrowDate32, isNullable: false)
+        return schemaBuilder.addField("col1", type: ArrowType(ArrowType.ArrowUInt8), isNullable: true)
+            .addField("col2", type: ArrowType(ArrowType.ArrowString), isNullable: false)
+            .addField("col3", type: ArrowType(ArrowType.ArrowDate32), isNullable: false)
             .finish()
     }
-
+
```
Comment (kou) [inline swift/Arrow/Tests/ArrowTests/IPCTests.swift:138]:
```suggestion

```

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift [hunk 11]
```diff
@@ -153,7 +154,7 @@ final class IPCFileReaderTests: XCTestCase {
         date32Builder.append(date2)
         date32Builder.append(date1)
         date32Builder.append(date2)
-
+
```
Comment (kou) [inline swift/Arrow/Tests/ArrowTests/IPCTests.swift:157]:
```suggestion

```

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift [hunk 12]
```diff
@@ -209,4 +212,163 @@ final class IPCFileReaderTests: XCTestCase {
...
+        uint8Builder.append(22)
+        uint8Builder.append(33)
+        uint8Builder.append(44)
+        let binaryBuilder = try ArrowArrayBuilders.loadBinaryArrayBuilder();
+        binaryBuilder.append("test10".data(using: .utf8))
+        binaryBuilder.append("test22".data(using: .utf8))
+        binaryBuilder.append("test33".data(using: .utf8))
+        binaryBuilder.append("test44".data(using: .utf8))
+        
+        let intHolder = ArrowArrayHolder(try uint8Builder.finish())
```
Comment (kou) [inline swift/Arrow/Tests/ArrowTests/IPCTests.swift:260]:
```suggestion
        let uintHolder = ArrowArrayHolder(try uint8Builder.finish())
```

BTW, do we need this column?

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift [hunk 13]
```diff
@@ -209,4 +212,163 @@ final class IPCFileReaderTests: XCTestCase {
...
+        time64Builder.append(1)
+        time64Builder.append(nil)
+        time64Builder.append(98765432)
+        let time32Builder = try ArrowArrayBuilders.loadTime32ArrayBuilder(.Milliseconds);
+        time32Builder.append(1)
+        time32Builder.append(2)
+        time32Builder.append(nil)
+        time32Builder.append(3)
+        
+        let intHolder = ArrowArrayHolder(try uint8Builder.finish())
```
Comment (kou) [inline swift/Arrow/Tests/ArrowTests/IPCTests.swift:297]:
```suggestion
        let uintHolder = ArrowArrayHolder(try uint8Builder.finish())
```

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift [hunk 14]
```diff
@@ -209,4 +212,163 @@ final class IPCFileReaderTests: XCTestCase {
...
+        let time64Holder = ArrowArrayHolder(try time64Builder.finish())
+        let time32Holder = ArrowArrayHolder(try time32Builder.finish())
+        let result = RecordBatch.Builder()
+            .addColumn("col1", arrowArray: intHolder)
+            .addColumn("col2", arrowArray: time64Holder)
+            .addColumn("col3", arrowArray: time32Holder)
```
Comment (kou) [inline swift/Arrow/Tests/ArrowTests/IPCTests.swift:303]: How about using more readable column names such as `time64` and  `time32`?

**Review by kou — APPROVED** [review]
+1

Could you use our pull request template next time?

Could you use 1 issue/PR per change for easy to review next time? For example, 3 issues/PRs for this case:

1. `ArrowType.info` -> `ArrowType` 
2. Add `binary`
3. Add `time32` and `time64`

**Comment by abandy** [comment]
> +1
> 
> Could you use our pull request template next time?
> 
> Could you use 1 issue/PR per change for easy to review next time? For example, 3 issues/PRs for this case:
> 
> 1. `ArrowType.info` -> `ArrowType`
> 2. Add `binary`
> 3. Add `time32` and `time64`

Yes, I will use the template and have 1 issue per PR going forward.  I apologize for any confusion.

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift (hunk 14 above)
Comment (abandy) [reply swift/Arrow/Tests/ArrowTests/IPCTests.swift:303]: Good idea, I will update.

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift (hunk 13 above)
Comment (abandy) [reply swift/Arrow/Tests/ArrowTests/IPCTests.swift:297]: Will change.

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift (hunk 12 above)
Comment (abandy) [reply swift/Arrow/Tests/ArrowTests/IPCTests.swift:260]: Will change.

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift (hunk 11 above)
Comment (abandy) [reply swift/Arrow/Tests/ArrowTests/IPCTests.swift:157]: Will remove.

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift (hunk 10 above)
Comment (abandy) [reply swift/Arrow/Tests/ArrowTests/IPCTests.swift:138]: Will remove.

File: swift/Arrow/Tests/ArrowTests/ArrayTests.swift (hunk 5 above)
Comment (abandy) [reply swift/Arrow/Tests/ArrowTests/ArrayTests.swift:156]: Will do.

File: swift/Arrow/Sources/Arrow/ArrowWriterHelper.swift (hunk 4 above)
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowWriterHelper.swift:47]: Will remove.

File: swift/Arrow/Sources/Arrow/ArrowType.swift (hunk 3 above)
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowType.swift:158]: Will remove.

File: swift/Arrow/Sources/Arrow/ArrowArray.swift (hunk 2 above)
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowArray.swift:227]: Will update.

File: swift/Arrow/Sources/Arrow/ArrowArray.swift (hunk 1 above)
Comment (abandy) [reply swift/Arrow/Sources/Arrow/ArrowArray.swift:198]: Will update.

File: swift/Arrow/Tests/ArrowTests/ArrayTests.swift (hunk 6 above)
Comment (abandy) [reply swift/Arrow/Tests/ArrowTests/ArrayTests.swift:172]: Will do.

Comment (abandy) [reply swift/Arrow/Tests/ArrowTests/ArrayTests.swift:172]: Will change.

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift (hunk 7 above)
Comment (abandy) [reply swift/Arrow/Tests/ArrowTests/IPCTests.swift:74]: Will remove.

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift (hunk 8 above)
Comment (abandy) [reply swift/Arrow/Tests/ArrowTests/IPCTests.swift:119]: Will remove.

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift (hunk 9 above)
Comment (abandy) [reply swift/Arrow/Tests/ArrowTests/IPCTests.swift:130]: Will remove.

File: swift/Arrow/Tests/ArrowTests/IPCTests.swift (hunk 12 above)
Comment (abandy) [reply swift/Arrow/Tests/ArrowTests/IPCTests.swift:260]: We don't I have been making all the test tables at least two columns.  But I will remove the unnecessary columns.

**Comment by abandy** [comment]
@kou I hope all is well.  Can you please merge this PR if everything looks ok?

**Comment by kou** [comment]
Merged!

**Comment by abandy** [comment]
Thank you!

**Pull request merged**